design/design-doc-composition-functions.md

```diff
@@ -175,32 +200,606 @@ metadata:
+
+```yaml
+apiVersion: pkg.crossplane.io/v1beta1
+kind: Function
+metadata:
+  name: go-templates
+spec:
+  package: xpkg.upbound.io/negz/go-templates:v0.1.0
+status:
+  # The gRPC endpoint where Crossplane will send RunFunctionRequests.
+  endpoint: https://go-templates-9sdfn2
+```
+
+gRPC communication between Crossplane and a Function will be secured by mutual
+transport layer security (mTLS). Crossplane has an established pattern for
+setting up mTLS certificate bundles - we do this today to secure communication
+between the API server and our validation webhooks, and between Crossplane and
+External Secret Stores. We might want to [rethink this][cert-per-entity] in
+future, but to begin with we'll stick with the established pattern, i.e.:
+
+* The Helm chart creates an empty Secret for the client bundle.
+* The core init container bootstraps the client bundle.
+* The package manager creates a Secret for each Function's server bundle.
+* The package manager bootstraps each Function's server bundle.
+
+See [#3884] for an example of this pattern.
+
+The frequency with which Crossplane will call each Function is relative to how
+frequently each XR is reconciled. There are two main triggers to reconcile an XR:
+
+1. A watch-triggered reconcile, i.e. because the XR changed.
+2. A poll-triggered reconcile, by default once per 60 seconds.
+
+XRs rely on poll-triggered reconciliation to promptly correct drift of their
+composed resources. The poll interval can be set using the `--poll-interval`
+flag. The XR controller does not know what kinds of resources it will compose
+when started, so it cannot start a watch for them.
+
+Assume:
+
+* There are 100 XRs, all of the same type, and all using the same Composition.
+* The XRs have all reached a steady state, polling at the default 60s frequency.
+* The Composition includes 1 Function call.
+
+In this case the Function would be called __100 * 1 = 100 times per 60s__.
+
+Most gRPC server implementations can serve multiple requests concurrently. For
+example in Go [each RPC is invoked in its own goroutine][go-grpc-concurrency].
+The gRPC project maintains a [suite of benchmarks][grpc-benchmarks]. At the time
+of writing these benchmarks show that an 8-core Python server can serve ~3,500
+trivial requests per second, while an 8-core Go server can serve ~180,000 per
+second. This should be considered a ceiling rather than indicative of real-world
+Function performance. In practice the throughput of a particular Function will
+be determined by the language it's written in, and what it does.
+
+Inevitably some Functions will be called frequently enough to exhaust the
+processing capacity of a single pod. A few potential approaches are possible to
+scale Functions include:
+
+* Tune the `--poll-interval`, e.g. to every 5m or 10m, to reduce calls.
+* Scale heavily used Functions horizontally, by increasing Deployment replicas.
+* Cache the responses of idempotent Function calls (see [Caching](#caching)).
```

Comment:
do we call out that function must be deterministic? (nit: idem-potency is something else).

Reply:
Actually, we need something like "eventually idem-potent", i.e. for every x some n with `fn^n(x) == x`, i.e. convergence.

Reply:
Good catch - will update. (I mix up these terms all the time - symptom of not having had a formal CS education.)